core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -213,15 +222,38 @@ public ShardId shardId(ClusterState clusterState, String index, String id, @Null
         return new ShardId(indexMetaData.getIndex(), generateShardId(indexMetaData, id, routing));
     }
 
-    static int generateShardId(IndexMetaData indexMetaData, String id, @Nullable String routing) {
-        final int hash;
-        if (routing == null) {
-            hash = Murmur3HashFunction.hash(id);
+    static int generateShardId(IndexMetaData indexMetaData, @Nullable String id, @Nullable String routing) {
+        if (indexMetaData.isRoutingPartitionedIndex()) {
+            if (routing == null) {
+                throw new IllegalArgumentException("expected a routing value on partitioned index [" + indexMetaData.getIndex().getName() + "]");
+            }
+
+            return generatePartitionedShardId(indexMetaData, routing, calculatePartitionOffset(indexMetaData, id));
```

Comment:
unless I miss something which is possible :) we could share a common codepath here since if `#getRoutingPartitionSize()` will be `1` in the unpartitioned case. This would simplify this method again and we can turn the check into an assertion that a routing value is provided? we would also get rid of most of the static onliners here (inline them into `generateShardId`). I think we should strive for a common code-path it's much easier to reason about and JIT might like it too